[CSS Fonts] Implement two-value syntax of font-size-adjust

shivamidow opened the pull request:
#### 2f42bab7265555901c27c9dbe5cef0b0c341ff14
<pre>
[CSS Fonts] Implement two-value syntax of font-size-adjust
<a href="https://bugs.webkit.org/show_bug.cgi?id=254191">https://bugs.webkit.org/show_bug.cgi?id=254191</a>

Reviewed by Myles C. Maxfield.

This change implements the two-value syntax of font-size-adjust specified in
the CSS Fonts Module Level 5 [1]. Five new font metrics: ex-height, cap-height
ch-width, ic-width and ic-height, are added, which clarify what font metric the
given value is for.

This change does not handle the CSS animation support and the &apos;from-font&apos; value.
Follow-up patches will cover them.

Test:
imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-009.html
imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-010.html
imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-011.html

[1] <a href="https://www.w3.org/TR/css-fonts-5/#font-size-adjust-prop">https://www.w3.org/TR/css-fonts-5/#font-size-adjust-prop</a>

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/incremental/stylesheet-body-incremental-rendering.html:
  Modified to deterministically measure repainted regions caused by style changes.
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/font-size-adjust-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
(WebCore::fontShorthandValue):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontSizeAdjust):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKeyRareData::create):
(WebCore::FontDescriptionKeyRareData::fontSizeAdjust const):
(WebCore::FontDescriptionKeyRareData::FontDescriptionKeyRareData):
(WebCore::FontDescriptionKey::FontDescriptionKey):
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialFontSizeAdjust):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::italic const):
(WebCore::FontDescription::fontSizeAdjust const):
(WebCore::FontDescription::setComputedSize):
(WebCore::FontDescription::setFontSizeAdjust):
(WebCore::FontDescription::decode):
(WebCore::FloatMarkableTraits::isEmptyValue): Deleted.
(WebCore::FloatMarkableTraits::emptyValue): Deleted.
* Source/WebCore/platform/graphics/FontPlatformData.cpp:
(WebCore::FontPlatformData::updateSizeWithFontSizeAdjust):
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/FontSizeAdjust.h: Added.
(WebCore::FloatMarkableTraits::isEmptyValue):
(WebCore::FloatMarkableTraits::emptyValue):
(WebCore::FontSizeAdjust::operator== const):
(WebCore::FontSizeAdjust::operator!= const):
(WebCore::add):
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setFontSizeAdjust):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::fontSizeAdjust const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontSizeAdjust):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontSizeAdjust):
* Source/WebCore/style/StyleFontSizeFunctions.cpp:
(WebCore::Style::adjustedFontSize):
* Source/WebCore/style/StyleFontSizeFunctions.h:

Canonical link: <a href="https://commits.webkit.org/262309@main">https://commits.webkit.org/262309@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52590718050a89a3542058a0a29ee73be48efbbd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1175 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1207 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1245 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1909 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1059 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1159 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1258 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1281 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1224 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1183 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1086 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1786 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1129 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1084 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1072 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1075 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1114 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2168 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1122 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1036 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1061 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1093 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/295 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1144 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4896,6 +4896,29 @@ static std::optional<Vector<FontFamilyRaw>> consumeFontFamilyRaw(CSSParserTokenR
+
+    RefPtr<CSSPrimitiveValue> value = consumeNumber(range, ValueRange::NonNegative);
+    if (value)
+        return value;
```
```suggestion
    if (auto value = consumeNumber(range, ValueRange::NonNegative))
        return value;
```

shivamidow replied:
Done.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4896,6 +4896,29 @@ static std::optional<Vector<FontFamilyRaw>> consumeFontFamilyRaw(CSSParserTokenR
+    if (value)
+        return value;
+
+    RefPtr<CSSPrimitiveValue> fontMetric = consumeIdent<CSSValueExHeight, CSSValueCapHeight, CSSValueChWidth, CSSValueIcWidth, CSSValueIcHeight>(range);
```
```suggestion
    auto fontMetric = consumeIdent<CSSValueExHeight, CSSValueCapHeight, CSSValueChWidth, CSSValueIcWidth, CSSValueIcHeight>(range);
```

shivamidow replied:
Done.

nt1m commented on Source/WebCore/platform/graphics/FontCascadeCache.h:
```diff
@@ -96,12 +96,12 @@ struct FontDescriptionKeyRareData : public RefCounted<FontDescriptionKeyRareData
 inline void add(Hasher& hasher, const FontDescriptionKeyRareData& key)
 {
-    add(hasher, key.featureSettings(), key.variationSettings(), key.variantAlternates(), key.fontPalette(), key.fontSizeAdjust());
+    add(hasher, key.featureSettings(), key.variationSettings(), key.variantAlternates(), key.fontPalette(), (*key.fontSizeAdjust()).value);
```
Does the hasher need to take in account the metric too?

litherum replied:
How does the hasher know the fontSizeAdjust isn't nullopt?

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4896,6 +4896,29 @@ static std::optional<Vector<FontFamilyRaw>> consumeFontFamilyRaw(CSSParserTokenR
+    if (!fontMetric)
+        return nullptr;
+
+    value = consumeNumber(range, ValueRange::NonNegative);
```
```suggestion
    auto value = consumeNumber(range, ValueRange::NonNegative);
```

shivamidow replied:
Done.

nt1m commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2506,12 +2506,13 @@ class PropertyWrapperFontSizeAdjust final : public PropertyWrapperGetter<std::op
-            return std::max(blendedAdjust, 0.0f);
+            auto blendedAdjust = blendFunc((*from.fontSizeAdjust()).value.value(), (*to.fontSizeAdjust()).value.value(), context);
+            FontSizeAdjust fontSizeAdjust = { (*from.fontSizeAdjust()).metric, std::max(blendedAdjust, 0.0f) };
+            return fontSizeAdjust;
```
Can you add a FIXME to properly interpolate between different metrics?

shivamidow replied:
Yes, added.

litherum commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -357,12 +357,17 @@ static RefPtr<CSSValue> valueForNinePieceImage(CSSPropertyID propertyID, const N
 }
 
-static Ref<CSSPrimitiveValue> fontSizeAdjustFromStyle(const RenderStyle& style)
+static Ref<CSSValue> fontSizeAdjustFromStyle(const RenderStyle& style)
```
It looks like every codepath through this function returns a CSSPrimitiveValue, so it looks like there's no need to make the return type more generic

shivamidow replied:
There was a missing piece. We return a CSSValuePair for FontSizeAdjust with non-ExHeight metrics.

litherum commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -172,8 +160,9 @@ class FontDescription {
     AtomString m_locale;
     AtomString m_specifiedLocale;
 
+    std::optional<FontSizeAdjust> m_sizeAdjust;
```
Shouldn't the optional be moved inside the FontSizeAdjust? We're not saying "there's optionally a font size adjust", but instead we're saying "there is always a font size adjust, and the value of it may be none"

So, perhaps something like `using FontSizeAdjust = std::optional<InnerFontSizeAdjust>`?

litherum commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1508,6 +1510,25 @@ inline FontVariationSettings BuilderConverter::convertFontVariationSettings(Buil
+
+    ASSERT(value.isPair());
+    const auto& pair = downcast<CSSValuePair>(value);
+    FontSizeAdjust fontSizeAdjust = { fromCSSValueID<FontSizeAdjust::Metric>(downcast<CSSPrimitiveValue>(pair.first()).valueID()), downcast<CSSPrimitiveValue>(pair.second()).floatValue() };
```
Nit: for clarity I'd prefer to pull out `.first()` and `.second()` onto previous lines and give them variable names.

Something like:
```
auto metric = fromCSSValueID<FontSizeAdjust::Metric>(downcast<CSSPrimitiveValue>(pair.first()).valueID());
auto value = downcast<CSSPrimitiveValue>(pair.second()).floatValue();
FontSizeAdjust fontSizeAdjust { WTFMove(metric), WTFMove(value) };
```

shivamidow replied:
O.K.

litherum commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1903,13 +1903,7 @@ inline void BuilderCustom::applyValueFontSize(BuilderState& builderState, CSSVal
-        ASSERT(primitiveValue.valueID() == CSSValueNone || CSSPropertyParserHelpers::isSystemFontShorthand(primitiveValue.valueID()));
-        fontDescription.setFontSizeAdjust(std::nullopt);
-    }
+    fontDescription.setFontSizeAdjust(BuilderConverter::convertFontSizeAdjust(builderState, value));
```
This is awesome

litherum commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -24,6 +24,7 @@
 
 #pragma once
 
+#include "FontSizeAdjust.h"
```
Can we forward declare `FontSizeAdjust` instead of including the header?

shivamidow replied:
Done

litherum commented on Source/WebCore/style/StyleFontSizeFunctions.cpp:
```diff
@@ -184,21 +184,32 @@ int legacyFontSizeForPixelSize(int pixelFontSize, bool shouldUseFixedDefaultSize
+        return metrics.hasCapHeight() ? adjustedFontSize(size, *sizeAdjust.value, metrics.floatCapHeight()) : size;
+    case FontSizeAdjust::Metric::ChWidth:
+        return metrics.zeroWidth() ? adjustedFontSize(size, *sizeAdjust.value, *metrics.zeroWidth()) : size;
+        break;
```
`break;` seems unnecessary

shivamidow replied:
Oops. Removed.

litherum commented on Source/WebCore/style/StyleFontSizeFunctions.cpp:
```diff
@@ -184,21 +184,32 @@ int legacyFontSizeForPixelSize(int pixelFontSize, bool shouldUseFixedDefaultSize
-        return size;
+    switch (sizeAdjust.metric) {
+    case FontSizeAdjust::Metric::CapHeight:
+        return metrics.hasCapHeight() ? adjustedFontSize(size, *sizeAdjust.value, metrics.floatCapHeight()) : size;
```
I think we should add a link to https://github.com/w3c/csswg-drafts/issues/6384 saying that this behavior isn't well-defined in the spec.

shivamidow replied:
Added.

litherum commented on Source/WebCore/style/StyleFontSizeFunctions.cpp:
```diff
@@ -184,21 +184,32 @@ int legacyFontSizeForPixelSize(int pixelFontSize, bool shouldUseFixedDefaultSize
+    case FontSizeAdjust::Metric::ChWidth:
+        return metrics.zeroWidth() ? adjustedFontSize(size, *sizeAdjust.value, *metrics.zeroWidth()) : size;
+        break;
+    case FontSizeAdjust::Metric::IcWidth:
```
This fallthrough doesn't seem right; we probably want to at least add a FIXME with a link to a bug about adding a real metric here. (Or to remove IcWidth from the spec)

shivamidow replied:
I added FIXME for further investigation here. When I tried, Gecko handled ic-height and ic-width in the same way, so I adopted the same approach.

litherum reviewed (approved):
r=me assuming EWS is green before landing

shivamidow reviewed (commented):
Thanks for your review. I didn't expect a formal review yet, but it was helpful. I will address your concerns and update the expected results soon.

FYI: In the next few updates, I will try CQ bots and see if they are happy, which are not for review.

vitorroriz commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2506,12 +2506,14 @@ class PropertyWrapperFontSizeAdjust final : public PropertyWrapperGetter<std::op
-            auto blendedAdjust = blendFunc(from.fontSizeAdjust().value(), to.fontSizeAdjust().value(), context);
-            return std::max(blendedAdjust, 0.0f);
+            auto blendedAdjust = blendFunc(*(*from.fontSizeAdjust()).value, *(*to.fontSizeAdjust()).value, context);
+            // FIXME: Handle interpolation between metrics.
```
```suggestion
            // FIXME: Handle interpolation between metrics (rdar://107049799).
```

It is linked to https://bugs.webkit.org/show_bug.cgi?id=254266

shivamidow replied:
Thanks.

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -357,12 +357,18 @@ static RefPtr<CSSValue> valueForNinePieceImage(CSSPropertyID propertyID, const N
+    FontSizeAdjust::Metric metric = (*fontSizeAdjust).metric;
+    float value = *(*fontSizeAdjust).value;
+    if (metric == FontSizeAdjust::Metric::ExHeight)
+        return CSSPrimitiveValue::create(value);
```
```suggestion
    auto metric = fontSizeAdjust->metric;
    float value = *fontSizeAdjust->value;
    if (metric == FontSizeAdjust::Metric::ExHeight)
        return CSSPrimitiveValue::create(value);
```

shivamidow reviewed (commented):
Now all EWS bots are green. \o/ If there are no other concerns, I will land this patch today and move on.

shivamidow committed:
[CSS Fonts] Implement two-value syntax of font-size-adjust
https://bugs.webkit.org/show_bug.cgi?id=254191

Reviewed by Myles C. Maxfield.

This change implements the two-value syntax of font-size-adjust specified in
the CSS Fonts Module Level 5 [1]. Five new font metrics: ex-height, cap-height
ch-width, ic-width and ic-height, are added, which clarify what font metric the
given value is for.

This change does not handle the CSS animation support and the 'from-font' value.
Follow-up patches will cover them.

Test:
imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-009.html
imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-010.html
imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-011.html

[1] https://www.w3.org/TR/css-fonts-5/#font-size-adjust-prop

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/incremental/stylesheet-body-incremental-rendering.html:
  Modified to deterministically measure repainted regions caused by style changes.
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/font-size-adjust-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-size-adjust-valid-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/layer-counter-style-override-expected.txt.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
(WebCore::fontShorthandValue):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontSizeAdjust):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKeyRareData::create):
(WebCore::FontDescriptionKeyRareData::fontSizeAdjust const):
(WebCore::FontDescriptionKeyRareData::FontDescriptionKeyRareData):
(WebCore::FontDescriptionKey::FontDescriptionKey):
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialFontSizeAdjust):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::italic const):
(WebCore::FontDescription::fontSizeAdjust const):
(WebCore::FontDescription::setComputedSize):
(WebCore::FontDescription::setFontSizeAdjust):
(WebCore::FontDescription::decode):
(WebCore::FloatMarkableTraits::isEmptyValue): Deleted.
(WebCore::FloatMarkableTraits::emptyValue): Deleted.
* Source/WebCore/platform/graphics/FontPlatformData.cpp:
(WebCore::FontPlatformData::updateSizeWithFontSizeAdjust):
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/FontSizeAdjust.h: Added.
(WebCore::FloatMarkableTraits::isEmptyValue):
(WebCore::FloatMarkableTraits::emptyValue):
(WebCore::FontSizeAdjust::operator== const):
(WebCore::FontSizeAdjust::operator!= const):
(WebCore::add):
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setFontSizeAdjust):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::fontSizeAdjust const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontSizeAdjust):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontSizeAdjust):
* Source/WebCore/style/StyleFontSizeFunctions.cpp:
(WebCore::Style::adjustedFontSize):
* Source/WebCore/style/StyleFontSizeFunctions.h:

Canonical link: https://commits.webkit.org/262309@main
